### ISSUE TITLE
Can't unload video

### ISSUE THREAD

**Issue by Luni-4** [bubblesub/bubblesub#60]
When I try to unload a video or an audio stream, I got the following traceback
```
problem running unload-audio:
[e] [12:35:07.038279] stream <bubblesub.api.audio_stream.AudioStream object at 0x7fc0eddc8690> is not loaded
[e] [12:35:07.039859] Traceback (most recent call last):
[e] [12:35:07.041447]   File "/usr/local/lib/python3.7/dist-packages/bubblesub/api/cmd.py", line 259, in run_async
[e] [12:35:07.042965]     await cmd.run()
[e] [12:35:07.044105]   File "/usr/local/lib/python3.7/dist-packages/bubblesub/cmd/file_ops.py", line 316, in run
[e] [12:35:07.045257]     self.api.audio.unload_current_stream()
[e] [12:35:07.046271]   File "/usr/local/lib/python3.7/dist-packages/bubblesub/api/threading.py", line 48, in _wrapper
[e] [12:35:07.047319]     return wrapped(*args, **kwargs)
[e] [12:35:07.048254]   File "/usr/local/lib/python3.7/dist-packages/bubblesub/api/base_streams_api.py", line 148, in unload_current_stream
[e] [12:35:07.049292]     self.unload_stream(self._current_stream.uid)
[e] [12:35:07.050345]   File "/usr/local/lib/python3.7/dist-packages/bubblesub/api/threading.py", line 48, in _wrapper
[e] [12:35:07.051346]     return wrapped(*args, **kwargs)
[e] [12:35:07.052261]   File "/usr/local/lib/python3.7/dist-packages/bubblesub/api/base_streams_api.py", line 120, in unload_stream
[e] [12:35:07.053193]     self.switch_stream(self._streams[index])
[e] [12:35:07.054107]   File "/usr/local/lib/python3.7/dist-packages/bubblesub/api/threading.py", line 48, in _wrapper
[e] [12:35:07.055027]     return wrapped(*args, **kwargs)
[e] [12:35:07.056743]   File "/usr/local/lib/python3.7/dist-packages/bubblesub/api/base_streams_api.py", line 141, in switch_stream
[e] [12:35:07.059486]     raise ValueError(f"stream {uid} is not loaded")
[e] [12:35:07.060322] ValueError: stream <bubblesub.api.audio_stream.AudioStream object at 0x7fc0eddc8690> is not loaded
```
Both the video and audio can still be reproducible inside `bubblesub` after the unload operation.

**Comment by rr-** [comment]
Thanks for this report. Pushed a bugfix on master, lmk if it works for you.

**Comment by Luni-4** [comment]
I can confirm that this issue is fixed. Thanks :)

**Comment by Luni-4** [comment]
It's not totally fixed, when I unload the video, the last frame is still rendered.

**Comment by rr-** [comment]
Can you interact with the video, though?

**Comment by Luni-4** [comment]
Nope, it's just a still image. Rendering the main window again fixes the problem though

**Comment by rr-** [comment]
Can't reproduce it: https://0x0.st/z53r.webm

**Comment by Luni-4** [comment]
Hmm, strange. It doesn't happen with audio in my case because it seems we force the spectrogram to render it again

**Comment by rr-** [comment]
@Luni-4 I pushed a bugfix on master that retracts the spectrogram cache when an audio stream is unloaded. Let me know if it's better.

**Comment by Luni-4** [comment]
@rr-  Thanks for the bugfix but the audio stream works now, the main problem is the video stream. The last frame viewed isn't removed from the video player, we need to render the video player again after having unloaded the video stream.

**Comment by rr-** [comment]
@Luni-4 since I can't reproduce it on my end I'd need your assistance; is this patch working for you?

```diff
diff --git a/bubblesub/ui/mpv.py b/bubblesub/ui/mpv.py
index 6e09a2f..2c9f8b5 100644
--- a/bubblesub/ui/mpv.py
+++ b/bubblesub/ui/mpv.py
@@ -129,6 +129,7 @@ class MpvWidget(QtWidgets.QOpenGLWidget):
     def _on_video_state_change(self, stream: VideoStream) -> None:
         self._sync_media()
         self._need_subs_refresh = True
+        self.update()
 
     def _on_audio_state_change(self, stream: AudioStream) -> None:
         self._sync_media()
```

**Comment by Luni-4** [comment]
Nope, it doesn't work. Looking at it much better, I think the video isn't unloaded at all because if I move forward the spectrogram, the frame label shows timestamps and frames numbers. While at it, the video player is always stuck to the last frame rendered before launching the unload command.

**Comment by rr-** [comment]
Super weird. What if you input `list-video` command?

**Comment by Luni-4** [comment]
Before launching the unload command:

`[i] [21:51:58.702093] X 0. 6b9c6209-084d-4fd6-af77-b06bc9789025 /my/path/to/video - video-name.mkv`

After the unload command:

`no loaded video streams`

If I don't pause the video and I launch the command, both the video player and the spectrogram aren't updated, the same goes for the frame label.

**Comment by rr-** [comment]
I suspect the problem is with

```
    def _sync_media(self) -> None:
        self._mpv.set_property("pause", True)
        self._mpv.command("loadfile", "null://")
        external_files: T.Set[str] = set()
        for stream in self._api.video.streams:
            external_files.add(str(stream.path))
        for stream in self._api.audio.streams:
            external_files.add(str(stream.path))
        self._mpv.set_property("external-files", list(external_files))
        if not external_files:
            self._api.playback.state = PlaybackFrontendState.NotReady
        else:
            self._api.playback.state = PlaybackFrontendState.Loading
```

but I can't really debug it myself

**Comment by Luni-4** [comment]
I will look into that personally soon or later

**Comment by Luni-4** [comment]
The `_sync_media` method works as expected, just checked.

I recorded a video to better explain what's happening. Looking at the cursor, you can see three different problems.
- Video unloaded but the last frame is still there
- The spectrogram cursor is split in two
- The frame label is still there

[https://0x0.st/zCAe.mkv](https://0x0.st/zCAe.mkv)

**Comment by Luni-4** [comment]
Thinking about this issue, would it be possible to add some tests to check video/audio stream loading/unloading? Since our tests run on `Ubuntu`, we could catch bug like this one immediately.

**Comment by rr-** [comment]
I agree such tests would be useful to have, but I'm not the one who's going to write them :—] Dealing with mpv and bubblesub AV pipeline is going to be rather tiresome.

**Comment by Luni-4** [comment]
I could write those tests but I need a plan first. I'm going to use a video sample added in the `data` folder. I've thought something like this:

Possible video test

1. Load video stream 
2. Run some video commands 
3. Unload video 
4. Run the same video commands above ---> if there are exceptions for all of them, it's correct. 

Same approach for audio streams.

**Comment by Luni-4** [comment]
It seems fixed in 95d105b
